src/query/storage/converter.go

```diff
@@ -76,49 +85,111 @@ func PromLabelsToM3Tags(
 // PromTimeSeriesToSeriesAttributes extracts the series info from a prometheus
 // timeseries.
 func PromTimeSeriesToSeriesAttributes(series prompb.TimeSeries) (ts.SeriesAttributes, error) {
+	switch series.Source {
+	case prompb.Source_PROMETHEUS:
+		return seriesAttributesForPrometheusSource(series)
+
+	case prompb.Source_OPEN_METRICS:
+		return seriesAttributesForOpenMetricsSource(series)
+
+	case prompb.Source_GRAPHITE:
+		return seriesAttributesForGraphiteSource(series)
+
+	default:
+		return ts.SeriesAttributes{}, fmt.Errorf("invalid source type %s", series.Source)
+	}
+}
+
+func seriesAttributesForPrometheusSource(series prompb.TimeSeries) (ts.SeriesAttributes, error) {
 	var (
-		sourceType        ts.SourceType
-		m3MetricType      ts.M3MetricType
 		promMetricType    ts.PromMetricType
 		handleValueResets bool
 	)
 
-	switch series.Source {
-	// TODO(linasm): implement internal support for OPEN_METRICS
-	case prompb.Source_PROMETHEUS, prompb.Source_OPEN_METRICS:
-		sourceType = ts.SourceTypePrometheus
-	case prompb.Source_GRAPHITE:
-		sourceType = ts.SourceTypeGraphite
+	switch series.Type {
+	case prompb.MetricType_UNKNOWN:
+		promMetricType = ts.PromMetricTypeUnknown
+
+	case prompb.MetricType_COUNTER:
+		promMetricType = ts.PromMetricTypeCounter
+		handleValueResets = true
+
+	case prompb.MetricType_GAUGE:
+		promMetricType = ts.PromMetricTypeGauge
+
+	case prompb.MetricType_HISTOGRAM:
+		promMetricType = ts.PromMetricTypeHistogram
+		handleValueResets = true
+
+	case prompb.MetricType_GAUGE_HISTOGRAM:
+		promMetricType = ts.PromMetricTypeGaugeHistogram
+		name := metricNameFromLabels(series.Labels)
+		handleValueResets = bytes.HasSuffix(name, promCountSuffix) ||
+			bytes.HasSuffix(name, openMetricsGaugeCountSuffix)
+
+	case prompb.MetricType_SUMMARY:
+		promMetricType = ts.PromMetricTypeSummary
+		name := metricNameFromLabels(series.Labels)
+		handleValueResets = bytes.HasSuffix(name, promCountSuffix) ||
+			bytes.HasSuffix(name, promSumSuffix)
+
+	case prompb.MetricType_INFO:
+		promMetricType = ts.PromMetricTypeInfo
+
+	case prompb.MetricType_STATESET:
+		promMetricType = ts.PromMetricTypeStateSet
+
 	default:
-		return ts.SeriesAttributes{}, fmt.Errorf("invalid source type %v", series.Source)
+		return ts.SeriesAttributes{}, fmt.Errorf("invalid metric type for Prometheus: %s", series.Type)
```

Comment:
Is it good that we are failing here? We would drop data in this case. Maybe we could set type to unknown and alert on that?

Reply:
I think this is not a problem because the protobuf is owned by the same component as this code..